File: common/src/main/java/dev/obscuria/fragmentum/v2/api/client/TooltipComponentRegistry.java
package dev.obscuria.fragmentum.v2.api.client;

import dev.obscuria.fragmentum.v2.core.client._TooltipComponentRegistry;
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent;
import net.minecraft.world.inventory.tooltip.TooltipComponent;

import java.util.Optional;
import java.util.function.Function;

@SuppressWarnings("unused")
public final class TooltipComponentRegistry {

    public static <T extends TooltipComponent> void register(Class<T> type, Function<T, ClientTooltipComponent> factory) {
        _TooltipComponentRegistry.register(type, factory);
    }

    public static Optional<ClientTooltipComponent> create(TooltipComponent component) {
        return _TooltipComponentRegistry.create(component);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/core/client/TooltipComponentRegistry.kt
package dev.obscuria.fragmentum.core.client

import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent
import net.minecraft.world.inventory.tooltip.TooltipComponent
import org.apache.commons.compress.utils.Lists
import java.util.*
import java.util.function.Function

object TooltipComponentRegistry
{
    private val definitions: MutableList<Definition<*>> = Lists.newArrayList()

    @JvmStatic
    fun <T : TooltipComponent> register(type: Class<T>, factory: Function<T, ClientTooltipComponent>)
    {
        if (definitions.stream().anyMatch { definition: Definition<*> -> definition.type == type }) return
        definitions.add(Definition(type, factory))
    }

    @JvmStatic
    fun create(component: TooltipComponent): Optional<ClientTooltipComponent>
    {
        for (definition in definitions)
            if (definition.isFor(component))
                return definition.create(component)
        return Optional.empty()
    }

    @JvmRecord
    private data class Definition<T : TooltipComponent>(
        val type: Class<T>,
        val factory: Function<T, ClientTooltipComponent>
    )
    {
        fun isFor(component: TooltipComponent): Boolean
        {
            return component.javaClass == type
        }

        fun create(component: TooltipComponent): Optional<ClientTooltipComponent>
        {
            return Optional.of(
                factory.apply(type.cast(component))
            )
        }
    }
}